[SPARK-2645] [Core] Allow SparkEnv.stop() to be called multiple times without side effects.

rekhajoshm opened the pull request:
Fix for SparkContext stop behavior - Allow sc.stop() to be called multiple times without side effects.

rekhajoshm committed:
Merge pull request #1 from apache/master

Pulling functionality from apache spark

rekhajoshm committed:
Merge pull request #2 from apache/master

pull latest from apache spark

rekhajoshm committed:
Merge pull request #3 from apache/master

Pulling functionality from apache spark

rekhajoshm committed:
SPARK-2645: Fix for SparkContext stop behavior

AmplabJenkins commented:
Can one of the admins verify this patch?

sarutak commented:
Jenkins, this is ok to test.

sarutak commented:
Hi @rekhajoshm , could you explain details more in the description field?

srowen commented:
Yeah, this might be a good change (`stop()` should be idempotent right?) but I'm not 100% clear how it connects to the JIRA. Also, there's no need to log an exception here.

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -90,17 +90,25 @@ class SparkEnv (
 
   private[spark] def stop() {
+
+    if(isStopped) return
```
might be better to rewrite this as
```
if (!isStopped) {
  // stop everything
}
```

srowen replied:
Heh, I would have gone with just 

```
if (isStopped) {
  return
}
```

to avoid the extra running indent but leave it.

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -90,17 +90,25 @@ class SparkEnv (
+      outputCommitCoordinator.stop()
+      rpcEnv.shutdown()
+    } catch {
+      case e: Exception =>
```
would be better to use `case NonFatal(e)` here

andrewor14 commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35709 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35709/consoleFull) for   PR 6973 at commit [`b566b66`](https://github.com/apache/spark/commit/b566b66dbf6dd380512f11fcee2ef77d0461522b).

SparkQA commented:
[Test build #35709 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35709/console) for   PR 6973 at commit [`b566b66`](https://github.com/apache/spark/commit/b566b66dbf6dd380512f11fcee2ef77d0461522b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Hi @rekhajoshm 

The PR look reasonable (with the suggested changes from andrew).  Could you also (a) change the title / description to be a more descriptive, eg. "Allow sc.stop() to be called multiple times" and also (b) add a test case to `SparkContextSuite` to prevent future regression?

thanks for working on this!

rekhajoshm committed:
SPARK-2645: Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35745 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35745/consoleFull) for   PR 6973 at commit [`380c5b0`](https://github.com/apache/spark/commit/380c5b00c8b76546afab68cc39cd1f648b298856).

SparkQA commented:
[Test build #35745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35745/console) for   PR 6973 at commit [`380c5b0`](https://github.com/apache/spark/commit/380c5b00c8b76546afab68cc39cd1f648b298856).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rekhajoshm committed:
SPARK-2645: Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35748 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35748/consoleFull) for   PR 6973 at commit [`58dba70`](https://github.com/apache/spark/commit/58dba70ac26fb77c1668a259a347603947787afe).

rekhajoshm commented:
Thanks for your quick inputs and review comments @sarutak @srowen @andrewor14 @squito .

@sean – Based on my quick look yesterday, sc.stop() calls sparkEnv stop() which tries to stop the already stopped and as it is not idempotent, causes a spiral effect of uncaught_exception of 50 exit code. https://issues.apache.org/jira/browse/SPARK-2645

@squito Did as suggested.Please review.Thanks

SparkQA commented:
[Test build #35748 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35748/console) for   PR 6973 at commit [`58dba70`](https://github.com/apache/spark/commit/58dba70ac26fb77c1668a259a347603947787afe).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -90,39 +92,47 @@ class SparkEnv (
+        outputCommitCoordinator.stop()
+        rpcEnv.shutdown()
+      } catch {
+        case NonFatal(e) =>
```
What's the value in adding this?

squito replied:
I think sean's point is that by catching the error, its no longer getting passed further up the stack.  If something unexpected goes wrong here, we most probably don't want to just log it, its likely to get overlooked and maybe indicates something seriously wrong.

srowen replied:
Yes, but further, I don't know if it helps to log it here either. I'd remove this `try` block.

andrewor14 replied:
I agree. Looks like the only thing this does is log the exception, but if we propagate it upwards it will be observed by the upper layers anyway. @rekhajoshm can you remove this `try` block?

srowen commented:
Jenkins retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

sujkh85 commented:
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-2645] [Core] Allow sc.stop() to be called multiple times without side effects. (#6973)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

SparkQA commented:
[Test build #35772 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35772/consoleFull) for   PR 6973 at commit [`58dba70`](https://github.com/apache/spark/commit/58dba70ac26fb77c1668a259a347603947787afe).

SparkQA commented:
[Test build #35772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35772/console) for   PR 6973 at commit [`58dba70`](https://github.com/apache/spark/commit/58dba70ac26fb77c1668a259a347603947787afe).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

squito commented on core/src/test/scala/org/apache/spark/SparkContextSuite.scala:
```diff
@@ -272,4 +273,24 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
+        // assert(!e.getMessage.contains("Server is already stopped"))
+        threwNoOrOnlyExceptedException = false
+      case NonFatal(e) =>
+        threwNoOrOnlyExceptedException = true
```
what exception are you expecting here?  I don't see any exception when running this test locally.

If you aren't expecting any exception, then you could just have the body of the `try` block in the test (any exception will fail the test).

If there is some "OK" exception, we should probably check more directly for that exception, as this will allow almost any exception to get thrown.

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rekhajoshm commented:
Thanks @srowen @squito for your review comments.
@srowen - log it for any kind of debugging, but can remove if not preferred
@squito - updated as suggested.

SparkQA commented:
[Test build #35806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35806/consoleFull) for   PR 6973 at commit [`9193a0c`](https://github.com/apache/spark/commit/9193a0cf315f07f149846cac973d245f832797e4).

squito commented:
hmm, now I am a little confused -- that test actually passes even without your changes.  Sorry this may have been my fault for suggesting a test in `SparkContextSuite` -- I guess `SparkContext` itself already handles multiple calls to `stop()`.

I'm not convinced this is still needed, but I suppose it doesn't hurt.  In any case, I'd leave the test you have, and also add a case where you just stop the env multiple times.

vanzin commented:
None of the changes here are in `SparkContext`; the PR title is misleading, since you're changing `SparkEnv`. What code path would cause `SparkEnv.stop()` to be called multiple times?

`SparkContext.stop()` is already idempotent:

    if (!stopped.compareAndSet(false, true)) {
      logInfo("SparkContext already stopped.")
      return
    }

SparkQA commented:
[Test build #35806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35806/console) for   PR 6973 at commit [`9193a0c`](https://github.com/apache/spark/commit/9193a0cf315f07f149846cac973d245f832797e4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35821/consoleFull) for   PR 6973 at commit [`a5a7d7f`](https://github.com/apache/spark/commit/a5a7d7f10693f06c51b57535e299236ebcfa37c0).

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rekhajoshm commented:
thanks @vanzin for your comments.Updated the title.Other than SparkContext, SparkEnv stop can be invoked from Scheduler, or if Driver commanded a shutdown, Executor stop would invoke SparkEnv stop, hence making SparkEnv idempotent as well.

On SPARK-2645, org.apache.spark.ServerStateException: Server is already stopped, from my understanding unhandled exception surfaces as uncaught_exception (exit code: 50)
Issue may be erratic, even be due to somewhere a async code state not updated correctly, shutDown hook, or (wild) scala compiler optimization within steps, or only in machine/usage specific variants :-)
From my very quick analysis, the one apparent culprit was stop() not being idempotent on SparkEnv.,  the fix was to prevent stopped env to be stopped again.

@squito @srowen good point; in enthusiasm to avoid unhandled exception especially when I am not expecting any or only in erratic maneuvers, but completely agree with you.updated.

Thanks again @vanzin @squito @srowen !

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #35821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35821/console) for   PR 6973 at commit [`a5a7d7f`](https://github.com/apache/spark/commit/a5a7d7f10693f06c51b57535e299236ebcfa37c0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented on core/src/test/scala/org/apache/spark/SparkContextSuite.scala:
```diff
@@ -272,4 +273,18 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
+      // call stop second time
+      sc.stop()
+    } catch {
+      case e: Exception =>
```
This `try` should be removed since it doesn't really add anything, except hiding the exception. Let it fly from the method to fail the test.

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35873/consoleFull) for   PR 6973 at commit [`12f66b5`](https://github.com/apache/spark/commit/12f66b52816152bc5a9db39964708e1c173c5286).

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35874/consoleFull) for   PR 6973 at commit [`1aff39c`](https://github.com/apache/spark/commit/1aff39c675a87f2d3f3be22a3955ddc2e4997d7b).

rekhajoshm commented:
thanks @srowen for your inputs.I had missed capturing exception on fail.updated(12f66b5).try/catch/fail(e) seems to be accepted practice when failing on exception.Used org.scalatest.Matchers "noException should be thrownBy" if that is preferred (1aff39c).Or please let me know if you have another preference.

```
import org.scalatest.Matchers._
noException should be thrownBy {
    sc = new SparkContext(new SparkConf().setAppName("test").setMaster("local"))
    val cnt = sc.parallelize(1 to 4).count()
    sc.cancelAllJobs()
    sc.stop()
    // call stop second time
    sc.stop()
  }
```

SparkQA commented:
[Test build #35874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35874/console) for   PR 6973 at commit [`1aff39c`](https://github.com/apache/spark/commit/1aff39c675a87f2d3f3be22a3955ddc2e4997d7b).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35875 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35875/consoleFull) for   PR 6973 at commit [`c97839a`](https://github.com/apache/spark/commit/c97839ace8b1f5407be0a4f4ed3202395b4c6e2f).

SparkQA commented:
[Test build #35873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35873/console) for   PR 6973 at commit [`12f66b5`](https://github.com/apache/spark/commit/12f66b52816152bc5a9db39964708e1c173c5286).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #35875 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35875/console) for   PR 6973 at commit [`c97839a`](https://github.com/apache/spark/commit/c97839ace8b1f5407be0a4f4ed3202395b4c6e2f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
@rekhajoshm I think this is still in your court. The extra try blocks should be removed IMHO.

rekhajoshm committed:
Fix for SparkContext stop behavior

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36022/consoleFull) for   PR 6973 at commit [`2ce5760`](https://github.com/apache/spark/commit/2ce57601e95623206db5acd38da09dc84a44044f).

SparkQA commented:
[Test build #36022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36022/console) for   PR 6973 at commit [`2ce5760`](https://github.com/apache/spark/commit/2ce57601e95623206db5acd38da09dc84a44044f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rekhajoshm commented:
@srowen done.can you please review/approve it?thanks.

andrewor14 commented:
@rekhajoshm LGTM other than a minor comment @srowen had.

rekhajoshm commented:
thanks @andrewor14 for your quick input.the last commit 2ce5760 was updated for the comment from @srowen kindly check.thanks

rekhajoshm committed:
Fix for SparkContext stop behavior

rekhajoshm commented:
had missed that inline comment. updated @andrewor14 @srowen thanks

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36077 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36077/consoleFull) for   PR 6973 at commit [`446b0a4`](https://github.com/apache/spark/commit/446b0a4206f07525b726413444b6354dca55fea3).

SparkQA commented:
[Test build #36077 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36077/console) for   PR 6973 at commit [`446b0a4`](https://github.com/apache/spark/commit/446b0a4206f07525b726413444b6354dca55fea3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -45,6 +45,8 @@ import org.apache.spark.storage._
 import org.apache.spark.unsafe.memory.{ExecutorMemoryManager, MemoryAllocator}
 import org.apache.spark.util.{RpcUtils, Utils}
 
+import scala.util.control.NonFatal
```
Not needed now?

srowen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -90,39 +92,44 @@ class SparkEnv (
+    if (!isStopped) {
+      isStopped = true
+
+      pythonWorkers.foreach { case (key, worker) => worker.stop()}
```
Very minor, but if you're changing the file again, add a space before the brace at the end, and this can be `{ case (_, worker) => ...`

andrewor14 replied:
or just `pythonWorkers.values.foreach(_.stop())`

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -90,39 +92,44 @@ class SparkEnv (
+      outputCommitCoordinator.stop()
+      rpcEnv.shutdown()
+
+
```
can you kill this blank line

rekhajoshm committed:
Fix for SparkContext stop behavior

rekhajoshm commented:
Thanks for your inputs @srowen @andrewor14. done.also removed unused JavaConversions._ please review, thanks.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36169/consoleFull) for   PR 6973 at commit [`277043e`](https://github.com/apache/spark/commit/277043eaac8f168da5d04715334fc7916afa577d).

SparkQA commented:
[Test build #36169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36169/console) for   PR 6973 at commit [`277043e`](https://github.com/apache/spark/commit/277043eaac8f168da5d04715334fc7916afa577d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Merging into master. Thanks for the fix @rekhajoshm

rekhajoshm commented:
thank you @andrewor14 @srowen for the reviews and merge.
